### ISSUE TITLE
OnRun()?

### ISSUE THREAD

**Issue by undeadindustries** [rivo/tview#136]
Is there a way to run a func when Run is called?

For example, my app needs to load many files into memory. I want it to start doing that before loading anything else, while displaying a spinner modal... I'd rather not have to have the user press or click anything for it to load these files.

I realize I could just load the files before .Run() but the user would just be waiting before any Box popped up...

Thanks!

**Comment by rivo** [comment]
I'm not completely sure I understand your request. Are you talking about the `Application`'s [`Run()`](https://godoc.org/github.com/rivo/tview#Application.Run) method? That method is typically only called once at the very beginning of your program. Nothing from `tview` is displayed before you call this function. You can run your function immediately before `Run()` or you can start it in parallel in a goroutine.

If you want to show some state during the loading of your files without allowing the user to interact with the application, you'll want to use [`Pages`](https://godoc.org/github.com/rivo/tview#Pages). Create a page with your state (e.g. A `TextView` with the text "Loading (80%)..."). When loading has finished, you switch to a page that contains the elements of your application.

If you want your application's elements to be visible already, you can also place a modal on top of them with the loading progress information. [Here](https://github.com/rivo/tview/wiki/Modal) is some documentation on how to work with modals.

Does this answer your question?

**Comment by undeadindustries** [comment]
Thank you for your reply.

Let me rephrase the question.

I need these files to open without the user having to press a button. I also want to load a spinner (page, modal, whatever) ...

So I was wondering how I could run a function when Focus() is run on an element. So the page would get focus and start loading these files.

Thanks!

**Comment by rivo** [comment]
I'm still not sure why you would need to hook into `Focus()`. I wrote up a quick example of how I understand your request and how I'd solve it:

```go
package main

import (
	"fmt"
	"time"

	"github.com/rivo/tview"
)

var app *tview.Application

func main() {
	pages := tview.NewPages()

	mainElements := tview.NewTextView().
		SetText("Files have been loaded").
		SetTextAlign(tview.AlignCenter)
	mainElements.SetTitle("Main application").
		SetBorder(true)
	pages.AddAndSwitchToPage("main", mainElements, true)

	app = tview.NewApplication()

	loadFiles(pages)

	if err := app.SetRoot(pages, true).Run(); err != nil {
		panic(err)
	}
}

func loadFiles(pages *tview.Pages) {
	progress := tview.NewTextView().SetText("Loading files...").SetTextAlign(tview.AlignCenter)
	progress.SetBorder(true)
	pages.AddAndSwitchToPage("loading", progress, true)

	go func() {
		// Simulate loading files with the time.Sleep() function.
		for file := 0; file < 10; file++ {
			time.Sleep(time.Second)
			fmt.Fprint(progress, ".")
			app.Draw()
		}
		pages.SwitchToPage("main")
		app.Draw()
		// We need to call app.Draw() because we're in a goroutine.
	}()
}
```

Does this do what you need?

**Comment by undeadindustries** [comment]
Thanks. That looks like it could work. Thanks!

**Comment by smazurov** [comment]
@rivo I just wanted to say thanks! Im jumping back into go to write a console app, and your project, responsiveness and code cleanliness is hugely appreciated. Can't thank you enough.

**Comment by rivo** [comment]
Thank you! I really appreciate it.

**Comment by mjgarton** [comment]
The example here is racy.  If you remove the sleep and try it with the race detector enabled, it fires fairly reliably.

Here is one example.

```

**Comment by rivo** [comment]
Thanks @mjgarton. I thought I had this one covered but I was wrong. `Draw()` is now thread-safe.

I found another race condition but it's part of [`tcell`](https://github.com/gdamore/tcell). I should probably submit an issue at some point.
